test/suites/color_diagnostics.bash

```diff
@@ -0,0 +1,125 @@
+if $COMPILER_TYPE_GCC ; then
+    color_diagnostics_enable='-fdiagnostics-color'
+    color_diagnostics_disable='-fno-diagnostics-color'
+elif $COMPILER_TYPE_CLANG ; then
+    color_diagnostics_enable='-fcolor-diagnostics'
+    color_diagnostics_disable='-fno-color-diagnostics'
+fi
+
+SUITE_color_diagnostics_PROBE() {
+    # Probe that real compiler actually supports colored diagnostics.
+    if [[ ! $color_diagnostics_enable || ! $color_diagnostics_disable ]] ; then
+        echo "compiler $COMPILER does not support colored diagnostics"
+    elif ! $REAL_COMPILER $color_diagnostics_enable -E - </dev/null >/dev/null 2>&1 ; then
+        echo "compiler $COMPILER (version: $compiler_version) does not support $color_diagnostics_enable"
+    elif ! $REAL_COMPILER $color_diagnostics_disable -E - </dev/null >/dev/null 2>&1 ; then
+        echo "compiler $COMPILER (version: $compiler_version) does not support $color_diagnostics_disable"
+    fi
+}
+
+SUITE_color_diagnostics_SETUP() {
+    if $run_second_cpp ; then
+        export CCACHE_CPP2=1
+    else
+        export CCACHE_NOCPP2=1
+    fi
+}
+
+color_diagnostics_expect_color() {
+    expect_file_contains "${1:?}" $'\033\\['
+    expect_file_contains <(fgrep 'previous prototype' "$1") $'\033\\['
+    expect_file_contains <(fgrep 'from preprocessor' "$1") $'\033\\['
```

Comment:
This won't produce a correct message in case of failure, as the process substitution can be read only once, so the "actual contents" message won't show anything.